### DISCUSSION TITLE
Issue with Viewing Last Line of Files Written on Local Device in MooseFS 3.0.116 Cluster

### DISCUSSION THREAD

**Discussion by sunshyer** [moosefs/moosefs#535] [Q&A]
I deployed a MooseFS 3.0.116 file system, with a total of 4 devices in the cluster. Each device has a separate process writing to individual files. However, I found that the last line of the file written by each device cannot be viewed on the current device, but can be viewed completely on other hosts within the cluster.

**Comment by chogata** [comment]
This is an attribute cache issue. The problem is, there is no method to just tell the kernel that the file length has changed. We have an option of invalidating ALL information about a file (that includes length, but also all other attributes and file data), but that is not very efficient. If we did it with every change to a file, it would be even less efficient than direct mode. We can also invalidate information about the "tail" of the file, but with the length increasing the problem is, if the kernel doesn't ask for new length, it doesn't know it should cache more data. And the kernel won't ask for new length unless attributes cache becomes invalid. Theoretically, the kernel should either invalidate or update file length after each write, but we've seen cases when this is not done properly.

In this situation you have a couple of solutions that might work:
1) Set just attribute cache to zero when mounting moosefs: `mount -o mfsattrcacheto=0 [...]`; this will make the kernel ask for file length (and other attributes) each time.
2) Or set this just for relevant files/directories using `mfsseteattr -f noattrcache OBJECT`
3) if 1 & 2 don't work as expected, you can try full on direct mode, again two ways to go about it:
- mount as direct: `mount -o mfscachemode=DIRECT [...]`
- set for relevant files/directories: `mfsseteattr -f nodatacache OBJECT`
